runtime/near-vm-runner/src/cache.rs

```diff
@@ -85,6 +87,110 @@ impl fmt::Debug for MockCompiledContractCache {
     }
 }
 
+struct FilesystemCompiledContractCacheState {
+    dir: rustix::fd::OwnedFd,
+    test_temp_dir: Option<tempfile::TempDir>,
+}
+
+#[derive(Clone)]
+pub struct FilesystemCompiledContractCache {
+    state: Arc<FilesystemCompiledContractCacheState>,
+}
+
+impl FilesystemCompiledContractCache {
+    pub fn new<HD: AsRef<std::path::Path> + ?Sized, SP: AsRef<std::path::Path> + ?Sized>(
+        home_dir: &HD,
+        store_path: Option<&SP>,
+    ) -> std::io::Result<Self> {
+        let path = std::env::var_os("NEAR_COMPILED_CACHE");
+        let path = path.map(PathBuf::from).unwrap_or_else(|| {
+            let store_path = store_path.map(AsRef::as_ref).unwrap_or_else(|| "data".as_ref());
+            home_dir.as_ref().join(store_path).join("contracts")
+        });
+        std::fs::create_dir_all(&path)?;
+        let dir = rustix::fs::open(path, rustix::fs::OFlags::DIRECTORY, rustix::fs::Mode::empty())?;
+        Ok(Self {
+            state: Arc::new(FilesystemCompiledContractCacheState { dir, test_temp_dir: None }),
+        })
+    }
+
+    pub fn test() -> std::io::Result<Self> {
+        let tempdir = tempfile::TempDir::new()?;
+        let mut cache = Self::new(tempdir.path(), None::<&str>)?;
+        Arc::get_mut(&mut cache.state).unwrap().test_temp_dir = Some(tempdir);
+        Ok(cache)
+    }
+}
+
+const ERROR_TAG: u8 = b'\n';
+const CODE_TAG: u8 = 0b10010101;
+/// Cache for compiled contracts code in plain filesystem.
+impl CompiledContractCache for FilesystemCompiledContractCache {
+    fn put(&self, key: &CryptoHash, value: CompiledContract) -> std::io::Result<()> {
+        use rustix::fs::{Mode, OFlags};
+        let final_filename = key.to_string();
+        let mut temp_file = tempfile::Builder::new().make_in("", |filename| {
+            let mode = Mode::RUSR | Mode::WUSR | Mode::RGRP | Mode::WGRP;
+            let flags = OFlags::CREATE | OFlags::TRUNC | OFlags::WRONLY;
+            Ok(std::fs::File::from(rustix::fs::openat(&self.state.dir, filename, flags, mode)?))
+        })?;
+        // This section manually "serializes" the data. The cache is quite sensitive to
+        // unnecessary overheads and in order to enable things like mmap-based file access, we want
+        // to have full control of what has been written.
+        match value {
+            CompiledContract::CompileModuleError(e) => {
```

Comment:
Sanity checking. Why do you need or want to store the error in cache? Is this to avoid compiling a erroneus contract over and over? Is there any other error condition that may give the same CompileModuleError but actually be recoverable in which case we may want to not put the error tag in cache and retry on the next attempt?

Reply:
> Why do you need or want to store the error in cache? Is this to avoid compiling a erroneus contract over and over?

This is attempting to be a non-functional change in this aspect, so even if there was a good reason to not store the error, it would be out of scope for this PR. But yes, we only need one attempt at (a potentially really expensive) compilation to figure out that the input is busted.

> Is there any other error condition that may give the same CompileModuleError but actually be recoverable in which case we may want to not put the error tag in cache and retry on the next attempt?

The compilation is intended to be deterministic as is the rest of the stack around this area of operation, so there isn't a good reason why the error would be recoverable. It may be, but that's more of an indication of a broader non-determinism problem, not an issue with the cache.